### ISSUE TITLE
Optionally format maps as JSON

### ISSUE THREAD

**Issue by mmarchini** [iovisor/bpftrace#445]
It will be easier for other tools to consume bpftrace output if we have an option to use a well known structured format. JSON is widely used and would fit well for our maps:

A few examples:

```
# bpftrace --json-maps -e 'tracepoint:raw_syscalls:sys_enter { @[comm] = count(); }'
Attaching 1 probe...
^C

{"@": {"bpftrace": 6, "systemd": 24, "snmp-pass": 96, "sshd": 125}}
```

```
bpftrace --json-maps -e 'kretprobe:vfs_read { @bytes = lhist(retval, 0, 2000, 200); }'
Attaching 1 probe...
^C

{"bytes": [ 
  {"min": 0, "max": 200, "count": 66},
  {"min": 200, "max": 400, "count": 2},
  {"min": 400, "max": 600, "count": 3},
  {"min": 800, "max": 1000, "count": 5},
  {"min": 2000, "count": 39} 
]}
```

We might need to put some thinking on more complex types for keys though, but we can start supporting primitive types for keys.

**Comment by brendangregg** [comment]
Makes sense, although perhaps just -j or --json. Also, using this option should force -q (quiet) to be on.

If the user calls time() with --json, we can print some JSON thing.

If the user calls printf() with --json, what should it do?

Also, how do we group things like timestamps with output? I could imagine this:

```
interval:s:1
{
    time();
    print(@lat);
}
```

Being automatically grouped by the action block. So the output would be:

```
{
  {"TIME": "8:23:01"}
  {"lat": [ 
    {"min": 0, "max": 200, "count": 66},
    {"min": 200, "max": 400, "count": 2},
    {"min": 400, "max": 600, "count": 3},
    {"min": 800, "max": 1000, "count": 5},
    {"min": 2000, "count": 39} 
  ]}
}
```

So any block that includes print/time/join statements gets an extra "{" "}" pair around it.

(replace TIME with however we want to format it in JSON).

**Comment by danobi** [comment]
I'm very interested in getting this feature in. I've been thinking about this for a bit and I have a couple thoughts.

> Being automatically grouped by the action block

I'm not sure that grouping by action block would be that useful. Usually I already know what output is going to come out of which probe. Extra separation might increase mental (and computational) overhead. 

Instead, I suggest having the bpftrace runtime batch output and throw each batch in a JSON object.  Each JSON object should then be framed by a newline to facilitate streamed parsing by another program. Eg:

```
{
  <stuff>
}

{
  <stuff>
}
```

> If the user calls printf() with --json, what should it do?

Each JSON object could have fields describing type, id, and output, where output is what would have otherwise printed to console.

**Comment by andreasgerstmayr** [comment]
Thanks @mmarchini for pointing me to this open issue. I started implementing this a few days back, my current WIP: #733

I started by moving the output parts to a `TextOutput` class, and moved shared functionality (e.g. `[l]hist_prepare`) into a base class `Output`.
In `JsonOutput` the JSON gets created in a streaming fashion.

Alternative way would be to create containers with the data and pass these to the text / json output functions, and include a JSON library for serialization.


Current output so far:
```
bpftrace -f json -e 'BEGIN { @scalar = 5; @complex[pid,comm] = 3; } tracepoint:raw_syscalls:sys_enter { @[comm] = count(); } kretprobe:vfs_read { @bytes = lhist(retval, 0, 2000, 200); } interval:s:1 { print(@scalar); print(@); print(@bytes); print(@complex); printf("\n\n"); }'
Attaching 4 probes...
{"@scalar": 5}

{"@": {"Curl Thread": 2, "AudioThread": 2, "sssd_be": 3, "systemd-journal": 5, "Socket Thread": 6, "Gecko_IOThread": 7, "Chrome_~dThread": 10, "JS Helper": 12, "JS Watchdog": 18, "WebExtensions": 22, "abrt-dump-journ": 24, "firefox": 30, "terminator": 32, "docker-containe": 42, "avahi-daemon": 42, "bpftrace": 110, "dockerd-current": 122, "PulseAudioDrive": 127, "polybar": 200, "Core Thread": 205, "compton": 236, "CompositorTileW": 334, "Compositor": 444, "Network Thread": 500, "alsa-sink-ALC25": 519, "Chrome_IOThread": 537, "Timer": 538, "Chrome_ChildIOT": 892, "Xorg": 1091, "threaded-ml": 1440, "pulseaudio": 1608, "Web Content": 2058, "spotify": 2957}}

{"@bytes": [
  {"max": -1, "count": 111}, 
  {"min": 0, "max": 199, "count": 1050}, 
  {"min": 200, "max": 399, "count": 0}, 
  {"min": 400, "max": 599, "count": 0}, 
  {"min": 600, "max": 799, "count": 0}, 
  {"min": 800, "max": 999, "count": 0}, 
  {"min": 1000, "max": 1199, "count": 0}, 
  {"min": 1200, "max": 1399, "count": 0}, 
  {"min": 1400, "max": 1599, "count": 0}, 
  {"min": 1600, "max": 1799, "count": 0}, 
  {"min": 1800, "max": 1999, "count": 0}, 
  {"min": 2000, "count": 1}
]}

{"@complex": {"4187,bpftrace": 3}}
```


**ToDos:**
* batching/grouping
* `printf`, other outputs
* multi-dimensional keys: JSON only allows strings as keys, so for now compound keys get joined by a comma (see `@complex` in the example).
* tests


> Instead, I suggest having the bpftrace runtime batch output and throw each batch in a JSON object. 

When does a batch start and stop? The print commands are received async in `perf_event_printer`. I could create a new `AsyncAction` which is fired when a block begins and ends?

**Comment by danobi** [comment]
> When does a batch start and stop? The print commands are received async in perf_event_printer. I could create a new AsyncAction which is fired when a block begins and ends?

I was imagining something like the runtime batching output every ~10ms or ~5 async prints, whichever comes first. I just looked at the code -- it looks like implementing that could get hairy. I think the way you have it (one json object per async print) is fine as a first pass. Maybe the docs should point out we reserve the right to start batching one day.

> ToDos:

One more thing: I think there should be type fields for each json async print. Otherwise the consumer will be stuck guessing how to interpret the value of each key.

**Comment by andreasgerstmayr** [comment]
I've implemented now everything except grouping in PR #733. If anybody comes up with a nice solution for grouping outputs of the same block I'd be happy to implement it (see my comment in the PR).

**Comment by danobi** [comment]
PR is merged
